source/server/guarddog_impl.cc

```diff
@@ -0,0 +1,103 @@
+#include "common/common/assert.h"
+#include "server/guarddog_impl.h"
+#include "server/watchdog_impl.h"
+
+namespace Server {
+
+GuardDogImpl::GuardDogImpl(Stats::Store& stats_store, const Server::Configuration::Main& config,
+                           SystemTimeSource& tsource)
+    : time_source_(tsource), miss_timeout_(config.wdMissTimeout()),
+      megamiss_timeout_(config.wdMegaMissTimeout()), kill_timeout_(config.wdKillTimeout()),
+      multi_kill_timeout_(config.wdMultiKillTimeout()),
+      loop_interval_([&]() -> std::chrono::milliseconds {
+        const auto min_of_nonfatal = std::min(miss_timeout_, megamiss_timeout_);
+        return std::min({killEnabled() ? kill_timeout_ : min_of_nonfatal,
+                         multikillEnabled() ? multi_kill_timeout_ : min_of_nonfatal,
+                         min_of_nonfatal});
+      }()),
+      watchdog_miss_counter_(stats_store.counter("server.watchdog_miss")),
+      watchdog_megamiss_counter_(stats_store.counter("server.watchdog_mega_miss")),
+      run_thread_(true) {
+  start();
+}
+
+GuardDogImpl::~GuardDogImpl() { exit(); }
```

Comment:
Oops, sorry, from quick read I thought this was exit(0) :)

Can we rename this function. :)